client/order/epochqueue.go

```diff
@@ -0,0 +1,165 @@
+func (eq *EpochQueue) Size() int {
+	eq.mtx.Lock()
+	size := len(eq.orders)
+	eq.mtx.Unlock()
+	return size
+}
+
+// Exists checks if the provided order id is in the queue.
+func (eq *EpochQueue) Exists(oid order.OrderID) bool {
+	eq.mtx.Lock()
+	_, ok := eq.orders[oid]
+	eq.mtx.Unlock()
+	return ok
+}
+
+// GenerateMatchProof calculates the sorting seed used in order matching
+// as well as the commitment checksum from the provided epoch queue
+// preimages and misses.
+//
+// The epoch queue needs to be reset if there are preimage mismatches or
+// non-existent orders for preimage errors.
+func (eq *EpochQueue) GenerateMatchProof(preimages []order.Preimage, misses []order.OrderID) (int64, msgjson.Bytes, error) {
+	eq.mtx.Lock()
+	defer eq.mtx.Unlock()
+
+	if len(eq.orders) == 0 {
+		return 0, nil,
+			fmt.Errorf("cannot generate match proof with an empty epoch queue")
+	}
+
+	// Remove all misses.
+	for _, oid := range misses {
+		delete(eq.orders, oid)
+	}
+
+	if len(eq.orders) != len(preimages) {
+		return 0, nil, fmt.Errorf("expected same length for epoch queue (%v) "+
+			"preimages (%v)", len(eq.orders), len(preimages))
+	}
+
+	// Map the preimages received with their associated epoch order ids.
+	matches := make(map[order.OrderID]order.Preimage, len(eq.orders))
+	for i := range preimages {
+		var match bool
+		for oid, commit := range eq.orders {
+			commitment := blake256.Sum256(preimages[i][:])
+			if bytes.Equal(commit[:], commitment[:]) {
+				matches[oid] = preimages[i]
+				match = true
+				break
+			}
+		}
+
+		if !match {
+			return 0, nil,
+				fmt.Errorf("no order match found for preimage %x", preimages[i])
+		}
+	}
+
+	// Ensure all remaining epoch orders matched to a preimage.
+	if len(matches) != len(eq.orders) {
+		return 0, nil, fmt.Errorf("expected all remaining epoch orders (%v) "+
+			"matched to a preimage (%v)", len(matches), len(eq.orders))
+	}
+
+	oids := make([]order.OrderID, 0, len(eq.orders))
+	commitments := make([]order.Commitment, 0, len(eq.orders))
+	for oid, commit := range eq.orders {
+		oids = append(oids, oid)
+		commitments = append(commitments, commit)
+	}
+
+	sort.Slice(oids, func(i, j int) bool {
+		return bytes.Compare(oids[i][:], oids[j][:]) < 0
+	})
+
+	sort.Slice(commitments, func(i, j int) bool {
+		return bytes.Compare(commitments[i][:], commitments[j][:]) < 0
+	})
+
+	// Concatenate the sorted preimages per and generate the
+	// seed.
+	sbuff := make([]byte, 0, len(eq.orders)*order.PreimageSize)
+	for _, oid := range oids {
+		pimg := matches[oid]
+		sbuff = append(sbuff, pimg[:]...)
+	}
+
+	sum := blake256.Sum256(sbuff)
+	seed := int64(binary.LittleEndian.Uint64(sum[:8]))
+
+	// Concatenate the sorted order commitments and generate the
+	// commitment checksum.
+	cbuff := make([]byte, 0, len(eq.orders)*order.CommitmentSize)
+	for _, commit := range commitments {
+		cbuff = append(cbuff, commit[:]...)
+	}
+	csum := blake256.Sum256(cbuff)
```

Comment:
We may consider creating a hasher to avoid actually concatenating into a large array, but using a new hasher ends up with more allocations, so this is OK for now.

i.e.
```go
hasher := blake256.New()
for i := range oids {
	pimg := orderPreimages[oids[i]]
	hasher.Write(pimg[:])
}
seed := hasher.Sum(nil)
```

Just something to consider if we don't want a very large `sbuff` and `cbuff` when order queues are large.